scipy/stats/tests/test_morestats.py

```diff
@@ -2975,90 +2981,96 @@ def test_directional_stats_correctness(self):
-        data = np.stack([np.cos(angles), np.sin(angles)], axis=1)
+    def test_directional_stats_2d_special_cases(self, angles, ref, xp):
+        angles = xp.asarray(angles)
+        ref = xp.asarray(ref)
+        data = xp.stack([xp.cos(angles), xp.sin(angles)], axis=1)
         res = 1 - stats.directional_stats(data).mean_resultant_length
-        assert_allclose(res, ref)
+        xp_assert_close(res, ref)
 
-    def test_directional_stats_2d(self):
+    def test_directional_stats_2d(self, xp):
         # Test that for circular data directional_stats
         # yields the same result as circmean/circvar
         rng = np.random.default_rng(0xec9a6899d5a2830e0d1af479dbe1fd0c)
-        testdata = 2 * np.pi * rng.random((1000, ))
-        testdata_vector = np.stack((np.cos(testdata),
-                                    np.sin(testdata)),
+        testdata = xp.asarray(2 * xp.pi * rng.random((1000, )))
+        testdata_vector = xp.stack((xp.cos(testdata),
+                                    xp.sin(testdata)),
                                    axis=1)
         dirstats = stats.directional_stats(testdata_vector)
         directional_mean = dirstats.mean_direction
-        directional_mean_angle = np.arctan2(directional_mean[1],
-                                            directional_mean[0])
-        directional_mean_angle = directional_mean_angle % (2*np.pi)
+        xp_test = array_namespace(directional_mean)  # np needs atan2
+        directional_mean_angle = xp_test.atan2(directional_mean[1],
+                                               directional_mean[0])
+        directional_mean_angle = directional_mean_angle % (2 * xp.pi)
         circmean = stats.circmean(testdata)
-        assert_allclose(circmean, directional_mean_angle)
+        xp_assert_close(directional_mean_angle, circmean)
 
-        directional_var = 1 - dirstats.mean_resultant_length
+        directional_var = 1. - dirstats.mean_resultant_length
         circular_var = stats.circvar(testdata)
-        assert_allclose(directional_var, circular_var)
+        xp_assert_close(directional_var, circular_var)
 
-    def test_directional_mean_higher_dim(self):
+    def test_directional_mean_higher_dim(self, xp):
         # test that directional_stats works for higher dimensions
         # here a 4D array is reduced over axis = 2
         data = np.array([[0.8660254, 0.5, 0.],
                          [0.8660254, -0.5, 0.]])
-        full_array = np.tile(data, (2, 2, 2, 1))
-        expected = np.array([[[1., 0., 0.],
-                              [1., 0., 0.]],
-                             [[1., 0., 0.],
-                              [1., 0., 0.]]])
+        full_array = xp.asarray(np.tile(data, (2, 2, 2, 1)))
+        expected = xp.asarray([[[1., 0., 0.],
+                                [1., 0., 0.]],
+                               [[1., 0., 0.],
+                                [1., 0., 0.]]],
+                              dtype=xp.float64)
```

Comment:
Should we perform this test with the default floating type ie
```suggestion
        full_array = xp.asarray(np.tile(data, (2, 2, 2, 1)).tolist())
        expected = xp.asarray([[[1., 0., 0.],
                                [1., 0., 0.]],
                               [[1., 0., 0.],
                                [1., 0., 0.]]])
```
?

Reply:
I'm happy with either, personally. I think there is value in doing some or most tests with default floating point type, but there is also some value in testing with non-default types. If that happens naturally by convenience in some tests, great.

Reply:
I'm only forcing the dtype of `expected`. For default torch input (float32), `directional_stats` returns `float64`. Is that okay?

Same thing for the next test function.

Reply:
Is that because of https://github.com/data-apis/array-api-compat/issues/152?

Reply:
I don't think so. I think it's just NumPy that converts to `float64`, and that's because of data-apis/array-api-compat#152 and the fact that `sum` is used. For `torch`, the dtype is preserved as it should be, yet strangely even if it is `float32`, `repr(f"{res.mean_resultant_length}")` prints lots of digits!

Reply:
What should we change. then? I assume you are referring to `xp.sum` being called within our call to `xp.mean`? We don't call `xp.sum` directly for PyTorch.

Reply:
Well, then, not even NumPy converts to `float64`. You wrote that "For default torch input (float32), directional_stats returns float64. Is that okay?" and I was trying to explain why that might be the case, but it doesn't look like that happens. I guess I didn't actually test for NumPy before, but it doesn't seem to happpen for NumPy either because - as you say - NumPy skips `sum`, so https://github.com/data-apis/array-api-compat/issues/152 doesn't come up.